### DISCUSSION TITLE
Grouped Allreduce not guarantee to group all allreduces tensors that user requests?

### DISCUSSION THREAD

**Discussion by zegao96** [horovod/horovod#3079] [Q&A]
Hi, 
I'm reading thru grouped allreduce feature that was introduced lately.  According to code here: https://github.com/horovod/horovod/blob/a76b0ffa93c409f447e5cf2a99e5133881c4dbcd/horovod/common/operations.cc#L1497 allreduces requested by one worker within the same group still send their requests separately during the master-slave negotiation process, so it appears to have no check or wait logic when the corresponding fused response is computed here: https://github.com/horovod/horovod/blob/a76b0ffa93c409f447e5cf2a99e5133881c4dbcd/horovod/common/controller.cc#L326
My question is what if some allreduce request in the group falls behind, does it end to be a partially grouped allreduce which is apparently not user-expected? Or I just took something wrong. 
One more thing to ask,  code here:
https://github.com/horovod/horovod/blob/a76b0ffa93c409f447e5cf2a99e5133881c4dbcd/horovod/common/controller.cc#L834
does it also end up with an incomplete grouped allreduce when the requested fused size exceeds HOROVOD_FUSION_THRESHOLD?
Looking forward to any sage insights on those doubts!
@romerojosh @tgaddair

**Comment by romerojosh** [comment]
Thanks for the questions @zgsbughammer!

On the enqueue side, the call to `process_set.tensor_queue.AddToTensorQueueMulti(entries, messages)` will take a lock on the tensor queue and add all the individual allreduce requests within a group *at the same time*. As a result, even though the requests within the group are still separate, they will become visible to the coordinator thread at the same cycle on a given worker, never split between multiple cycles. Since all workers follow this same logic, the requests in a group will never show up as globally ready by all workers and ready for reduction in split cycles either.

For your last question concerning `HOROVOD_FUSION_THRESHOLD` and incomplete groups: yes, if `HOROVOD_FUSION_THRESHOLD` size is smaller than the aggregated size of the tensors you are attempting to group together, while the requests will be processed for allreduce on the same cycle, the group will be split into more than one allreduce, launched back-to-back. To avoid this, ensure `HOROVOD_FUSION_THRESHOLD` is set large enough to contain the largest group you expect. 

Hope this helps!

**Reply by zegao96** [reply]
oh I see, code here:
https://github.com/horovod/horovod/blob/a76b0ffa93c409f447e5cf2a99e5133881c4dbcd/horovod/common/controller.cc#L426
shows that each worker would pop all the request messages from the tensor queue in the same cycle and then send the encoded version to the coordinator thread. It explains!
thanks for clearing the doubts! @romerojosh
